pkg/storage/stores/shipper/uploads/table_manager.go

```diff
@@ -0,0 +1,230 @@
+package uploads
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/local"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	pkg_util "github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type Config struct {
+	Uploader       string
+	IndexDir       string
+	UploadInterval time.Duration
+}
+
+type TableManager struct {
+	cfg             Config
+	boltIndexClient BoltDBIndexClient
+	storageClient   StorageClient
+
+	metrics   *metrics
+	tables    map[string]*Table
+	tablesMtx sync.RWMutex
+
+	done chan struct{}
+	wg   sync.WaitGroup
+}
+
+func NewTableManager(cfg Config, boltIndexClient BoltDBIndexClient, storageClient StorageClient, registerer prometheus.Registerer) (*TableManager, error) {
+	tm := TableManager{
+		cfg:             cfg,
+		boltIndexClient: boltIndexClient,
+		storageClient:   storageClient,
+		metrics:         newMetrics(registerer),
+		done:            make(chan struct{}),
+	}
+
+	tables, err := tm.loadTables()
+	if err != nil {
+		return nil, err
+	}
+
+	tm.tables = tables
+	return &tm, nil
+}
+
+func (tm *TableManager) loop() {
+	defer tm.wg.Done()
+
+	syncTicker := time.NewTicker(tm.cfg.UploadInterval)
+	defer syncTicker.Stop()
+
+	for {
+		select {
+		case <-syncTicker.C:
+			err := tm.uploadTables(context.Background())
+			if err != nil {
+				level.Error(pkg_util.Logger).Log("msg", "error uploading local boltdb files to the storage", "err", err)
+			}
+		case <-tm.done:
+			return
+		}
+	}
+}
+
+func (tm *TableManager) Stop() {
+	close(tm.done)
+	tm.wg.Wait()
+
+	err := tm.uploadTables(context.Background())
+	if err != nil {
+		level.Error(pkg_util.Logger).Log("msg", "error uploading local boltdb files to the storage before stopping", "err", err)
+	}
+}
+
+func (tm *TableManager) QueryPages(ctx context.Context, queries []chunk.IndexQuery, callback chunk_util.Callback) error {
+	return chunk_util.DoParallelQueries(ctx, tm.query, queries, callback)
+}
+
+func (tm *TableManager) query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	tm.tablesMtx.RLock()
+	defer tm.tablesMtx.RUnlock()
+
+	table, ok := tm.tables[query.TableName]
+	if !ok {
+		return nil
+	}
+
+	return table.Query(ctx, query, callback)
+}
+
+func (tm *TableManager) BatchWrite(ctx context.Context, batch chunk.WriteBatch) error {
+	boltWriteBatch, ok := batch.(*local.BoltWriteBatch)
+	if !ok {
+		return errors.New("invalid write batch")
+	}
+
+	for tableName, tableWrites := range boltWriteBatch.Writes {
+		table, err := tm.getOrCreateTable(tableName)
+		if err != nil {
+			return err
+		}
+
+		err = table.Write(ctx, tableWrites)
```

Comment:
Again this is just a question should this be done in parallel ?

Reply:
Same as above.